crates/accelerate/src/sparse_pauli_op.rs

```diff
@@ -60,8 +62,202 @@ pub fn unordered_unique(py: Python, array: PyReadonlyArray2<u16>) -> (PyObject,
+                        z[[i, j]].write(false);
+                        x[[i, j]].write(true);
+                    }
+                    Pauli::Y => {
+                        z[[i, j]].write(true);
+                        x[[i, j]].write(true);
+                        phase = phase.wrapping_add(1);
+                    }
+                    Pauli::Z => {
+                        z[[i, j]].write(true);
+                        x[[i, j]].write(false);
+                    }
+                }
+            }
+            phases[i].write(phase % 4);
+        }
+        // These are safe because the above loops write into every element.  It's guaranteed that
+        // each of the elements of the `paulis` vec will have `num_qubits` because they're all
+        // reading from the same base array.
+        let z = unsafe { z.assume_init() };
+        let x = unsafe { x.assume_init() };
+        let phases = unsafe { phases.assume_init() };
+        Ok(ZXPaulis {
+            z: z.into_pyarray(py).into(),
+            x: x.into_pyarray(py).into(),
+            phases: phases.into_pyarray(py).into(),
+            coeffs: PyArray1::from_vec(py, coeffs).into(),
+        })
+    }
+}
+
+/// Recurse worker routine of `decompose_dense`.  Should be called with at least one qubit.
+fn decompose_dense_inner(
+    factor: Complex64,
+    num_qubits: usize,
+    paulis: &[Pauli],
+    block: ArrayView2<Complex64>,
+    out_paulis: &mut Vec<Vec<Pauli>>,
+    out_coeffs: &mut Vec<Complex64>,
+    square_tolerance: f64,
+) {
+    if num_qubits == 0 {
+        // It would be safe to `return` here, but if it's unreachable then LLVM is allowed to
+        // optimise out this branch entirely in release mode, which is good for a ~2% speedup.
+        unreachable!("should not call this with an empty operator")
+    }
+    // Base recursion case.
+    if num_qubits == 1 {
+        let mut push_if_nonzero = |extra: Pauli, value: Complex64| {
+            if value.norm_sqr() <= square_tolerance {
+                return;
+            }
+            let paulis = {
+                let mut vec = Vec::with_capacity(paulis.len() + 1);
+                vec.extend_from_slice(paulis);
+                vec.push(extra);
+                vec
+            };
+            out_paulis.push(paulis);
+            out_coeffs.push(value);
+        };
+        push_if_nonzero(Pauli::I, 0.5 * factor * (block[[0, 0]] + block[[1, 1]]));
+        push_if_nonzero(Pauli::X, 0.5 * factor * (block[[0, 1]] + block[[1, 0]]));
+        push_if_nonzero(
+            Pauli::Y,
+            0.5 * Complex64::i() * factor * (block[[0, 1]] - block[[1, 0]]),
+        );
+        push_if_nonzero(Pauli::Z, 0.5 * factor * (block[[0, 0]] - block[[1, 1]]));
+        return;
+    }
+    let mut recurse_if_nonzero = |extra: Pauli, factor: Complex64, values: Array2<Complex64>| {
+        let mut is_zero = true;
+        for value in values.iter() {
+            if *value != Complex64::new(0.0, 0.0) {
```

Comment:
```suggestion
            if !value.is_zero() {
```
just to avoid making a new complex on each iteration.

Although, I wonder if we want to do `if !value.is_zero() || !value.is_subnormal()` instead. There is a `is_normal` function, but I'm not sure how we would handle a `NaN` or infinity here. I don't think it could ever come up so it probably doesn't matter.

Reply:
`is_normal` would definitely be the wrong check - zero and one are both normal while infinity is not, but an infinity probably should be propagated through (it definitely hasn't meant that the term is zero).  I think the only way `NaN` could appear without being passed in would be if we did `Infinity - Infinity`, but in that case, I think it's safer to return it than to silently suppress it.

`is_subnormal()` is an interesting call, but I'm not sure it's entirely correct for the logic here.  I'm not certain what logic you were meaning because your expression is true for all floats (the condition is the same as `! (value.is_zero() && value.is_subnormal())`, and nothing that `is_zero` is also subnormal).  If it was `!value.is_zero() && !value.is_subnormal()`, I'm not convinced it's completely correct logically - the `tolerance` argument can already be used to filter out subnormals in the final step (you can even set that to some subnormal yourself, if you like).  As an example, let's say we're at the final decomposition stage, we've got a matrix that looks like `[[a, a], [a, a]]`, the factor is 1, and `a` is subnormal.  The I- and X-like decompositions end up as `(a + a) * 0.5` which ends up being the valid, non-zero `a` in all cases, which should then be subject to the `tolerance` argument.  The purpose of this highlighted line is to filter out recursive components that can _never_ match the tolerance arg, but subnormals still can, and are totally valid floats.

The `is_zero` form is clearer, it just requires pulling in `num_traits` which we ought to do anyway - this isn't the first time I've wanted it.  Done in 8a82510005.

---

As a philosophical point, I'm not really sure how correct it is to say we're making a "new" `Complex64` after optimisations, since
- the type's `Copy`, small in size and stack-allocated, so it's not actually going to have much effect
- LLVM would happily keep the value around in a register through each loop iteration if it had space and deemed it necessary, and if it didn't, it'd have to load/store each time regardless of what we did
- the comparison to two constant floating-point zeros gets optimised to a SIMD "is zero?" comparison anyway.

But that's just unimportant musings about compilers.